### ISSUE TITLE
Hang when partitioning an empty region

### ISSUE THREAD

**Issue by bandokihiro** [StanfordLegion/legion#1064]
I experienced a hang which I was able to solve by avoiding partitiong an empty region by field. The partition is a second-level partition by field of a first partition by pre-image. According to @streichler, this is technically allowed. This happened with 6 ranks/sub-regions. The hang did not happen with 5 ranks/sub-regions which also contained a partition of an empty region.

The command line used is:
```
mpirun -n 6 -npernode 1 -bind-to none solver -logfile logs/log_%.log -hdf5:forcerw -dm:memoize -ll:csize 20000 -ll:rsize 0 -ll:cpu 0 -ll:util 2 -ll:onuma 0 -ll:ocpu 1 -ll:othr 14 -ll:force_kthreads -lg:inorder
```
I am on 9136995065015629002bd5d340d50b11726b7170.
Attached are the backtraces:
[gdb0.txt](https://github.com/StanfordLegion/legion/files/6393004/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/6393005/gdb1.txt)
[gdb2.txt](https://github.com/StanfordLegion/legion/files/6393006/gdb2.txt)
[gdb3.txt](https://github.com/StanfordLegion/legion/files/6393008/gdb3.txt)
[gdb4.txt](https://github.com/StanfordLegion/legion/files/6393009/gdb4.txt)
[gdb5.txt](https://github.com/StanfordLegion/legion/files/6393011/gdb5.txt)

**Comment by lightsighter** [comment]
> I experienced a hang which I was able to solve by avoiding partitiong an empty region by field

I don't think this has anything to do with the hang. Assuming this is the same hang that I looked at earlier this week, the issue has to do with deferred allocation and deletion. There were deferred allocations that Realm had handed out and never satisfied. The fact that that happened with `-lg:inorder` means it's highly liked to be Realm's fault because each operation in Legion will run from start to finish through the pipeline before the next operation is even allowed to start.

Please dump all the Realm logs for deferred allocation and deletion. @streichler can tell you the loggers and appropriate logging level.

**Comment by bandokihiro** [comment]
As an additional comment, this is not the same set-up as the one I showed earlier this week, not saying that the issue is not the same.

**Comment by lightsighter** [comment]
Please make me frozen processes on sapling with `-DDEBUG_LEGION` and legion spy logs.

**Comment by bandokihiro** [comment]
Location of logs: `/home/bandokihiro/Runs/TemporalMixingLayer2D/debug/logs`
I had to kill the job because the log file on rank 0 was growing unboundedly

**Comment by lightsighter** [comment]
That suggest that the job was not actually hung. Looks like a repeated series of user event creations and triggers. You should be able to attach to that process and print out some backtraces from the thread(s) that are changing over time.

**Comment by bandokihiro** [comment]
Wouldn't that be similar to the backtraces I originally posted? I will regenerate those once sapling is available.

**Comment by bandokihiro** [comment]
Since sapling is being used, I ran it on another cluster and took backtraces for the rank for which the log file was growing at 3 different times.

The command line is
```
mpirun -report-bindings -n 6 -npernode 1 -bind-to none solver -logfile logs/log_%.log -hdf5:forcerw -dm:memoize -ll:csize 20000 -ll:rsize 0 -ll:cpu 0 -ll:util 2 -ll:onuma 0 -ll:ocpu 1 -ll:othr 14 -lg:partcheck -lg:inorder -ll:force_kthreads -level spy=2
```

[gdb0.txt](https://github.com/StanfordLegion/legion/files/6415684/gdb0.txt)
[gdb1.txt](https://github.com/StanfordLegion/legion/files/6415685/gdb1.txt)
[gdb2.txt](https://github.com/StanfordLegion/legion/files/6415687/gdb2.txt)

**Comment by lightsighter** [comment]
Sapling doesn't appear busy to me. The CPU nodes aren't even close to being fully loaded. Your command line also does not work.

```
mebauer@c0001:/home/bandokihiro/Builds_CPUNodes/DG-Legion/build_master/exec$ mpirun -n 6 --bind-to none -H c0001,c0002,c0003,c0004 --oversubscribe /home/bandokihiro/Builds_CPUNodes/DG-Legion/build_master/exec/solver -logfile /home/mebauer/fuckme/logs/log_%.log -hdf5:forcerw -dm:memoize -ll:csize 20000 -ll:rsize 0 -ll:cpu 0 -ll:util 2 -ll:onuma 0 -ll:ocpu 1 -ll:othr 14 -lg:partcheck -lg:inorder -ll:force_kthreads -level spy=2
terminate called after throwing an instance of 'std::runtime_error'
  what():  toml::parse: file open error -> input.toml
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
terminate called after throwing an instance of 'std::runtime_error'
  what():  toml::parse: file open error -> input.toml
terminate called after throwing an instance of 'std::runtime_error'
terminate called after throwing an instance of 'std::runtime_error'
  what():  toml::parse: file open error -> input.toml
terminate called after throwing an instance of 'std::runtime_error'
  what():  toml::parse: file open error -> input.toml
  what():  toml::parse: file open error -> input.toml
terminate called after throwing an instance of 'std::runtime_error'
  what():  toml::parse: file open error -> input.toml
```

**Comment by bandokihiro** [comment]
I thought the new way of running jobs on sapling was to use slurm. According to sinfo, the nodes are all used so I didn't want to perturb someone else's run. Please add `-input /home/bandokihiro/Runs/TemporalMixingLayer2D/debug/input.toml` to the command line and it should work.

**Comment by lightsighter** [comment]
Pull and try again.

**Comment by bandokihiro** [comment]
Fixed in 7bc7708d.
